jscomp/core/lam_compile.ml

```diff
@@ -762,18 +762,23 @@ and compile_untagged_cases ~cxt ~switch_exp ~default ~block_cases cases =
     in
     E.emit_check check
   in
-  let is_array (l, _) = l = Ast_untagged_variants.Untagged ArrayType in
+  let is_not_typeof (l, _) = match l with 
+  | Ast_untagged_variants.Untagged (InstanceType _) -> true
+  | _ -> false in
   let switch ?default ?declaration e clauses =
-    let array_clauses = Ext_list.filter clauses is_array in
-    match array_clauses with
-    | [(l, {J.switch_body})] when List.length clauses > 1 ->
-      let rest = Ext_list.filter clauses (fun c -> not (is_array c)) in
+    let not_typeof_clauses = Ext_list.filter clauses is_not_typeof in
+    let typeof_clauses = Ext_list.filter clauses (fun c -> not (is_not_typeof c)) in
+    let rec build_if_chain remaining_clauses = (match remaining_clauses with 
+    | (Ast_untagged_variants.Untagged (InstanceType Array), {J.switch_body}) :: rest -> 
       S.if_ (E.is_array e)
         (switch_body)
-        ~else_:([S.string_switch ?default ?declaration (E.typeof e) rest])
-    | _ :: _ :: _ -> assert false (* at most 1 array case *)
-    | _ ->
-      S.string_switch ?default ?declaration (E.typeof e) clauses in
+        ~else_:([build_if_chain rest])
+    | (Ast_untagged_variants.Untagged (InstanceType instanceType), {J.switch_body}) :: rest -> 
+      S.if_ (E.instanceof e (E.js_global (Ast_untagged_variants.Instance.to_string instanceType)))
```

Comment:
See js_exp_make: the array case emits different code.
Perhaps  put the emission logic in one place shared so it's done only once for instances (and the array special cased only once in that logic)

Reply:
@cristianoc can you check this commit, is this along the lines of what you had in mind? https://github.com/rescript-lang/rescript-compiler/pull/6383/commits/e687695144bd75ec48f0b1eadaccd70dc84e1e0a

Logic is now only in `emit_check`.